### PR TITLE
Data movement

### LINKED ISSUE #29: Data Movement - Extract data
Extract the data from the RDS (SQL Server).

- [ ] Extract the data from the RDS
- [ ] Store the data in an intermediary parquet file (**confer with team to make sure this is still the current approach**).

Extracting the data from the RDS is a necessary step to put the data into the desired format for long-term storage.

### PR DESCRIPTION
**Pull request by Lasped13**
Wrote the script which extracts the readings from the last 24 hours in the database and stores them all locally into a parquet file.

Closes #29

### CODE REVIEW

File: data_movement/extract_from_rds.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+# pylint: disable=no-member
+
+from os import environ as ENV
+from datetime import datetime, timedelta
+import pandas as pd
+from dotenv import load_dotenv
+import pymssql
+
+
+def get_connection() -> object:
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:17]:
```suggestion
def get_connection(None) -> object:
```

File: data_movement/extract_from_rds.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        host=ENV["DB_HOST"],
+        port=int(ENV["DB_PORT"]),
+        database=ENV["DB_NAME"],
+        user=ENV["DB_USER"],
+        password=ENV["DB_PASSWORD"]
+    )
+    return conn
+
+
+def get_the_time() -> str:
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:29]:
```suggestion
def get_the_time(None) -> str:
```

File: data_movement/extract_from_rds.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    """This function gets the time 24 hours ago."""
+    now = datetime.now()
+    yesterday = now - timedelta(hours=24)
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:32]: Consider pulling out `24` into a global constant. Or even better if passed as a parameter.

File: data_movement/extract_from_rds.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    yesterday_str = yesterday.strftime('%Y-%m-%d %H:%M:%S')
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:33]: Put this pattern into a global contant.

File: data_movement/extract_from_rds.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        port=int(ENV["DB_PORT"]),
+        database=ENV["DB_NAME"],
+        user=ENV["DB_USER"],
+        password=ENV["DB_PASSWORD"]
+    )
+    return conn
+
+
+def get_the_time() -> str:
+    """This function gets the time 24 hours ago."""
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:30]:
```suggestion
    """This function gets the time 24 hours ago as a formatted string."""
```

File: data_movement/extract_from_rds.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    now = datetime.now()
+    yesterday = now - timedelta(hours=24)
+    yesterday_str = yesterday.strftime('%Y-%m-%d %H:%M:%S')
+
+    return yesterday_str
+
+
+def query_and_extract_rds(time: str, conn: object):
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:38]: Need to add a return type to type hint.

File: data_movement/extract_from_rds.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+SELECT *
+FROM alpha.FACT_plant_reading
+WHERE taken_at >= '{time}'
+""")
+        data = cur.fetchall()
+
+    return data
+
+
+def load_into_parquet(data: list):
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:51]:
```suggestion
def load_into_parquet(data: list) -> None:
```

File: data_movement/extract_from_rds.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    df = pd.DataFrame(data=data, columns=[
+                      "reading_id", "soil_moisture", "temperature", "last_watered",
+                      "taken_at", "plant_id", "botanist_id"])
+    df["plant_id"] = pd.to_numeric(
+        df["plant_id"], errors='coerce').astype('Int64')
+
+    df.to_parquet('readings_last_24_hours.parquet', index=False)
+
+
+def extract_data():
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:62]:
```suggestion
def extract_data(None) -> None:
```

File: data_movement/extract_from_rds.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+def load_into_parquet(data: list):
+    """This function loads the given data into a parquet file"""
+    df = pd.DataFrame(data=data, columns=[
+                      "reading_id", "soil_moisture", "temperature", "last_watered",
+                      "taken_at", "plant_id", "botanist_id"])
+    df["plant_id"] = pd.to_numeric(
+        df["plant_id"], errors='coerce').astype('Int64')
+
+    df.to_parquet('readings_last_24_hours.parquet', index=False)
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:59]: Put this into the .env and pass it into the function

File: data_movement/extract_from_rds.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        data = cur.fetchall()
+
+    return data
+
+
+def load_into_parquet(data: list):
+    """This function loads the given data into a parquet file"""
+    df = pd.DataFrame(data=data, columns=[
+                      "reading_id", "soil_moisture", "temperature", "last_watered",
+                      "taken_at", "plant_id", "botanist_id"])
```
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:55]: Consider putting these as a global constant. Or even better if passed into the function

**Review by BerkayDur — CHANGES_REQUESTED** [review]
Just a few changes and suggestions but very good code overall

File: data_movement/extract_from_rds.py (hunk 2 above)
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:29]: I also think this function could be tested.

File: data_movement/extract_from_rds.py (hunk 6 above)
Comment (BerkayDur) [inline data_movement/extract_from_rds.py:38]: Could test this function to check how many times it calls functions within in

**Review by BerkayDur — CHANGES_REQUESTED** [review]
Think about adding some testing

File: data_movement/extract.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+
+def query_and_extract_rds(time: str, conn: Connection) -> list[list]:
+    """This function creates the query to get the data start from time"""
+    with conn.cursor() as cur:
+        cur.execute(f"""
+SELECT *
+FROM alpha.FACT_plant_reading
+WHERE taken_at >= '{time}'
+""")
```
Comment (joe1606) [inline data_movement/extract.py:52]:
```suggestion
        cur.execute(f"""
        SELECT *
        FROM alpha.FACT_plant_reading
        WHERE taken_at >= '{time}'
        """)
```

Comment (joe1606) [reply data_movement/extract.py:52]: its bugging me, but if it interferes with tests then ignore it

**Review by joe1606 — DISMISSED** [review]
looks good

**Review by joe1606 — APPROVED** [review]
looks all good

**Pull request merged**